pydantic/_internal/_json_schema_shared.py

```diff
@@ -0,0 +1,92 @@
+"""
+Types and utility functions used by various other internal tools.
+"""
+from __future__ import annotations
+
+from typing import Any, Callable, Dict, Union
+
+from pydantic_core import core_schema
+
+JsonSchemaValue = Dict[str, Any]
+CoreSchemaOrField = Union[core_schema.CoreSchema, core_schema.DataclassField, core_schema.TypedDictField]
+
+
+class GetJsonSchemaHandler:
+    """
+    Handler to call into the next JSON schema generation function
```

Comment:
I realize this is re-exported and intended to be public, but I think if I were to jump-to-definition and got brought to this file I might feel like I lost my bearings a bit and wonder why I'm in an internal file.

Because of that, I think it would be helpful to give this class a little bit more detailed of a docstring, basically describing how it is meant to be used (and possibly with some example usage?), and reinforcing that it is meant for "users" to use.

Happy to help draft something for this, I think we should update it a bit before merging so we don't forget, but it doesn't have to be "final draft" quality.

Reply:
Maybe I'm misunderstanding how this should be used, I didn't realize it was an abstract class.

If it doesn't need to be abstract I think it might be better to not have it be this way as I think it increases the learning curve. (By that I mean, delete this class and change the name of `UnpackedRefJsonSchemaHandler` to the name this class currently has, then people can just subclass and override. I think that would help with discovering the implementation, etc.)

Or maybe I'm misunderstanding something about how this is meant to work?

Reply:
Okay I see you do use two different subclasses in internal code so maybe it's more confusing to give this a default implementation.

In that case, I think we should direct people to the two implementations (and possibly move the json schema one into this file? Not sure if that's possible.)

The main reason being I think it will be annoying for people to jump to the implementation while trying to debug code or whatever. (I personally hate when I am told that something matches an interface and then I can't jump to the definition, but if necessary, I'd hope we could at least make it easier to jump to the definition I care about.)

Reply:
Moved them all into a single file.